constraint/pkg/client/client.go

```diff
@@ -236,59 +270,121 @@ func (c *Client) CreateCRD(ctx context.Context, templ *templates.ConstraintTempl
+
+	if err := rewriteModulePackage(entryPointPath, entryPoint); err != nil {
+		return nil, err
+	}
 
 	req := ruleArities{
 		"violation": 1,
 	}
-	if err := requireRules(path, src, req); err != nil {
+	if err := requireRulesModule(entryPoint, req); err != nil {
 		return nil, fmt.Errorf("Invalid rego: %s", err)
 	}
 
-	_, err = c.rConformer.ensureRegoConformance(crd.Spec.Names.Kind, path, src)
+	rr.AddEntryPointModule(entryPointPath, entryPoint)
+	for idx, libSrc := range targetSpec.Libs {
+		libPath := fmt.Sprintf(`%s["lib_%d"]`, libPrefix, idx)
+		if err := rr.AddLib(libPath, libSrc); err != nil {
+			return nil, err
+		}
+	}
+
+	sources, err := rr.Rewrite()
 	if err != nil {
 		return nil, err
 	}
 
-	return crd, nil
+	var mods []constraintTemplateModule
+	if err := sources.ForEachModule(func(m *regorewriter.Module) error {
+		content, err := m.Content()
+		if err != nil {
+			return err
+		}
+		mods = append(mods, constraintTemplateModule{
+			path: m.Module.Package.Path.String(),
+			rego: string(content),
+		})
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return &constraintTemplateArtifacts{
+		crd:           crd,
+		targetHandler: targetHandler,
+		modules:       mods,
+	}, nil
+}
+
+// CreateCRD creates a CRD from template
+func (c *Client) CreateCRD(ctx context.Context, templ *templates.ConstraintTemplate) (*apiextensions.CustomResourceDefinition, error) {
+	artifacts, err := c.createTemplateArtifacts(templ)
+	if err != nil {
+		return nil, err
+	}
+	return artifacts.crd, nil
 }
 
 // AddTemplate adds the template source code to OPA and registers the CRD with the client for
 // schema validation on calls to AddConstraint. It also returns a copy of the CRD describing
 // the constraint.
 func (c *Client) AddTemplate(ctx context.Context, templ *templates.ConstraintTemplate) (*types.Responses, error) {
 	resp := types.NewResponses()
-	crd, err := c.CreateCRD(ctx, templ)
-	if err != nil {
-		return resp, err
-	}
-
-	var src string
-	var target TargetHandler
-	for _, v := range templ.Spec.Targets {
-		k := v.Target
-		t, ok := c.targets[k]
-		if !ok {
-			return resp, fmt.Errorf("Target %s not recognized", k)
-		}
-		target = t
-		src = v.Rego
-	}
 
-	path := createTemplatePath(target.GetName(), crd.Spec.Names.Kind)
-	conformingSrc, err := c.rConformer.ensureRegoConformance(crd.Spec.Names.Kind, path, src)
+	artifacts, err := c.createTemplateArtifacts(templ)
 	if err != nil {
 		return resp, err
 	}
 
 	c.constraintsMux.Lock()
 	defer c.constraintsMux.Unlock()
-	if err := c.backend.driver.PutModule(ctx, path, conformingSrc); err != nil {
-		return resp, err
+
+	var rollbackPaths []string
+	defer func() {
+		for _, path := range rollbackPaths {
+			if _, err := c.backend.driver.DeleteModule(ctx, path); err != nil {
+				glog.Warningf("encountered error while rolling back module %s", path)
+			}
+		}
+	}()
+
+	for _, mod := range artifacts.modules {
+		if err := c.backend.driver.PutModule(ctx, mod.path, mod.rego); err != nil {
```

Comment:
Currently PutModule triggers a full recompile each time. This is problematic here because:

1. The compile might fail if there are missing library functions the current module relies on
2. Uses CPU unnecessarily

I think the right thing to do here is to add a `PutModules()` function that is atomic. We can do this with the local driver no problem. @tsandall has mentioned that they were interested in adding this to OPA's rest API. In the interim the remote driver's `PutModules()` could just be a loop around `PutModule()`. This would make it impossible to use the remote driver with Kubernetes until the atomic update is fixed (not a blocker as K8s doesn't currently have the option to run remote OPA anyway.). I believe Torin and I had discussed this limitation and are fine with it (but cc-ing to raise awareness).

Reply:
re: point 1 - I'll add PutModules/DeleteModules (more detail in other comment).
re: point 2 - this is the same as the status quo, so I think we are conflating things if we want to address this here.

Reply:
+1 to both responses

Reply:
re: point 2: Sorry, I was thinking that point 2 was about avoiding recompilation on update. Current status quo is to recompile at most once per AddTemplate(), we should try to preserve that, as recompile can be expensive, and I'm assuming many libraries per constraint.

Reply:
I'm not sure I see where it's currently doing the recompile check, is that inside of PutModule()?

Reply:
I believe this is addressed via `PutModules()`